File: THF_SpringBoot/src/main/java/ProjetJDR/THF_SpringBoot/services/PersonnageService.java
package ProjetJDR.THF_SpringBoot.services;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import ProjetJDR.THF_SpringBoot.entity.Metiers;
import ProjetJDR.THF_SpringBoot.entity.Personnage;
import ProjetJDR.THF_SpringBoot.exceptions.PersonnageException;
import ProjetJDR.THF_SpringBoot.repositories.PersonnageRepository;

@Service
public class PersonnageService {
	@Autowired
	private PersonnageRepository personnageRepository;

	public List<Personnage> getAll() {
		return personnageRepository.findAll();
	}

	public Personnage getById(Long id) {
		return personnageRepository.findById(id).orElseThrow(() -> {
			throw new PersonnageException("Veuillez insérer un Id existant");
		});
	}

	public Personnage getByPrenom(String prenom) {
		return personnageRepository.findByPrenom(prenom).orElseThrow(() -> {
			throw new PersonnageException("Prénom introuvable");
		});
	}

	public Personnage getByNom(String nom) {
		return personnageRepository.findByNom(nom).orElseThrow(() -> {
			throw new PersonnageException("Nom introuvable");
		});

	}

	public Personnage getByIdWithStuff(Long id) {
		return personnageRepository.findByIdWithStuff(id).orElseThrow(() -> {
			throw new PersonnageException("Veuillez insérer un Id existant");
		});

	}

	// Retourne tous les personnages dont le metier est celui indiqué par
	// l'utilisateur
	public Personnage getPersonnageByMetier(Metiers metier) {
		return personnageRepository.findPersonnageByMetier(metier).orElseThrow(() -> {
			throw new PersonnageException("Veuillez insérer un Id existant");
		});

	}

	public Personnage create(Personnage personnage) {
		if (personnage.getNom() == null || personnage.getNom().isEmpty()) {
			throw new PersonnageException();
		}
		return personnageRepository.save(personnage);
	}

	public Personnage update(Personnage personnage) {
		return personnageRepository.save(personnage);
	}

	public void delete(Personnage personnage) {
		personnageRepository.delete(personnage);
	}

	public void deleteById(Long id) {
		personnageRepository.deleteById(id); // A voir --> suppression des attributs du personnage
	}

	public void deleteByNom(String nom) {
		personnageRepository.deleteByNom(nom);
	}

	public void deleteByPrenom(String prenom) {
		personnageRepository.deleteByPrenom(prenom);
	}
}

File: heroes-foundry-app/src/app/services/personnage.service.ts
import { Personnage } from '../model/personnage';
import { Metier } from './../model/metier';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class PersonnageService {

  url : string = 'http://localhost:8080/personnage'

  constructor(private httpClient: HttpClient) {}

  getById(id : number) : Observable<Personnage>{
    return this.httpClient.get<Personnage>(`${this.url}/${id}`);
    }

    getByPrenom(prenom : string ) : Observable<Personnage>{
      return this.httpClient.get<Personnage>(`${this.url}/${personnage.prenom}`, personnage);
      }

      getByNom(nom : string ) : Observable<Personnage>{
        return this.httpClient.get<Personnage>(`${this.url}/${personnage.nom}`, personnage);
  }

  getByIdWithStuff(id : number) : Observable<Personnage>{
    return this.httpClient.get<Personnage>(`${this.url}/${id}/${personnage.stuff}`, personnage)

}

// Retourne tous les personnages dont le metier est celui indiqué par
// l'utilisateur
getPersonnageByMetier(personnage : Personnage) : Observable<Personnage[]>{
    return this.httpClient.get<Personnage[]>(`${this.url}/${personnage.metier}}`, personnage)
}

//OU

getPersonnageByMetier(metier : Metier) : Observable<Personnage[]>{
  return this.httpClient.get<Personnage[]>(`${this.url}/${personnage.metier}}`, personnage)
}


create(personnage : Personnage) {
  let personnageJson = {
    nom : personnage.nom,
    prenom : personnage.prenom,
  };
  return this.httpClient.post<Personnage>(this.url, personnageJson);
}

update(personnage: Personnage): Observable<Personnage> {
  return this.httpClient.put<Personnage>(`${this.url}/${personnage.id}`, personnage);
}

delete(personnage : Personnage) : Observable<void> {
    return this.httpClient.delete<void>(`${this.url}/${personnage.id}`, personnage);
}

deleteById(id : number) : Observable<void> {
  return this.httpClient.delete<void>(`${this.url}/${id}`); // A voir --> suppression des attributs du personnage
}

deleteByNom(nom : string) : Observable<void> {
  return this.httpClient.delete<void>(`${this.url}/${nom}`);
}

deleteByPrenom(prenom : string) : Observable<void> {
  return this.httpClient.delete<void>(`${this.url}/${prenom}`);
}


}
